File: kconc.cpp
#include<bits/stdc++.h>
using namespace std;
#define lld long long int
lld maxSubArraySum(lld a[], lld size)
{
    lld max_so_far = INT_MIN, max_ending_here = 0;

    for (lld i = 0; i < size; i++)
    {
        max_ending_here = max_ending_here + a[i];
        if (max_so_far < max_ending_here)
            max_so_far = max_ending_here;

        if (max_ending_here < 0)
            max_ending_here = 0;
    }
    return max_so_far;
}
int main(){
  lld t;
  cin>>t;
  for(lld j=0;j<t;j++){
    lld n,k;
    cin>>n>>k;
    lld a[2*n];
    lld b[2*n];
    lld sum = 0;
    for(lld i=0; i<n; i++){
      cin>>a[i];
      sum+=a[i];
    }
    for(lld i=0;i<2*n;i++){
      b[i]=a[i%n];
    }
    lld y = maxSubArraySum(a,n);
    if(k==1){
      cout<<y<<endl;
    }
    else if(k==2){
      lld x = maxSubArraySum(b,2*n);
      cout<<x<<endl;
    }
    else{
      if(sum>=0){
        a[n] = sum*(k-2);
        for(int i=0;i<n;i++){
          a[i+n+1] = a[i];
        }
        lld x = maxSubArraySum(a,2*n+1);
        cout<<x<<endl;
      }
      else{
        lld x = maxSubArraySum(b,2*n);
        cout<<x<<endl;
      }
    }
  }
  return 0;
}

File: kconc.py
def maxSubArraySum(a,size):

    max_so_far =a[0]
    curr_max = a[0]

    for i in range(1,size):
        curr_max = max(a[i], curr_max + a[i])
        max_so_far = max(max_so_far,curr_max)

    return max_so_far
t = input()
while(t>0):
    t-=1
    n,k = map(int,raw_input().split())
    l = map(int,raw_input().split())
    y = maxSubArraySum(l,n)
    if k==1:
        print y
    elif k==2:
        b = []
        for i in range(2*n):
            b.append(l[i%n])
        x = maxSubArraySum(b,2*n)
        print x
    else:
        if sum(l)>=0:
            l.append(sum(l)*(k-2))
            for i in range(n):
                l.append(l[i])
            print maxSubArraySum(l,2*n+1)
        else:
            for i in range(n):
                l.append(l[i])
            print maxSubArraySum(l,2*n)
